### PR TITLE
fix(SHRINKRES-299): Support "Maven CI Friendly Versions"

### PR DESCRIPTION
**Pull request by famod**
by preferring the output of flatten-maven-plugin over the regular pom.xml
which just has a "useless" placeholder like ${revision} as the version.

**Some notes about this approach:**

There is already a TODO note in `ClasspathWorkspaceReader.createFoundArtifact(File)` about whether it would be better to load the pom (and its parents) "properly" via Maven.
This might have also been a possible solution, but I am not sure whether such a heavy lifting should be done there.
The note regarding "cycle in graph reconstruction" also scared me off a bit, to be honest.
Furthermore, a `revision` property does not need to defined in the pom at all (AFAIK). It can also be passed in via `-Drevision=...`. Parsing the entire Maven model would not help in that case.

My first approach was to completely swap the `FileInfo`:
```java
    private FileInfo createPomFileInfo(final File childFile) {

        // assuming that directory entry on classpath is target/classes directory, we need
        // to go two directories up in the structure and grab a pom.xml file from there
        // SHRINKRES-299: we prefer a "flattened" pom.xml written by flatten-maven-plugin
        File parent = childFile.getParentFile();
        if (parent != null) {
            parent = parent.getParentFile();
            if (parent != null) {
                FileInfo fileInfo = new FileInfo(new File(parent, flattenedPomPath));
                if (!fileInfo.isFile()) {
                    fileInfo = new FileInfo(new File(parent, "pom.xml"));
                }
                return fileInfo;
            }
        }

        return null;
    }
```
but this caused problems in `MavenResolvedArtifactImpl.artifactToFile(Artifact)` as the name of the artifact file would then not be `pom.xml` anymore but `.flattened-pom.xml` or something else. I would have needed to re-evaluate the new system property there. Or maybe `org.eclipse.aether.artifact.Artifact.getProperty(String, String)` could have been used (`ClasspathWorkspaceReader` might be able to set a property that signals "this artifact is a pom!").
I dropped that approach because it did not feel right.

And why did I introduce a new system property?
The name an the output directory of the flattened pom are completely configurable, see: https://www.mojohaus.org/flatten-maven-plugin/flatten-mojo.html#flattenedPomFilename
Also there's a ticket about changing the default output folder/name: https://github.com/mojohaus/flatten-maven-plugin/issues/53
So, with the system property people can adapt `shrinkwrap-resolver` to their flatten config.

### CODE REVIEW

**Comment by famod** [comment]
@bartoszmajsak I'll try to implement a test for this but as of now there is (apparently) no in-depth test for this class, only a "shallow" one testing dangling directories (`ClassPathScanningTestCase.classpathWithDanglingDirs()`).

**Comment by famod** [comment]
Test implemented. Actually I am pretty pleased with the outcome.

**Comment by famod** [comment]
This should be it, now. I had to fix a test failure by adding a normally ignored file and some typos.

**Comment by famod** [comment]
@bartoszmajsak Ready when you are! :-)

**Comment by bartoszmajsak** [comment]
> @bartoszmajsak Ready when you are! :-)

On it :)

File: maven/impl-maven/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/aether/ClassPathScanningTestCase.java [hunk 1]
```diff
@@ -25,38 +25,41 @@
...
 /**
+ * Tests {@link ClasspathWorkspaceReader}.
  *
  * @author <a href="kpiwko@redhat.com">Karel Piwko</a>
- *
+ * @author <a href="https://github.com/famod">Falko Modler</a>
+ * @since SHRINKRES-178
  */
-// SHRINKRES-178
 public class ClassPathScanningTestCase {
```
Comment (bartoszmajsak) [inline maven/impl-maven/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/aether/ClassPathScanningTestCase.java]: Super happy to see this test case being expanded to cover more corner cases. Kudos!

As we are dealing with system properties here - what do you think about using `@Rule` for that? I think it could simplify the test code a bit. In particular, I am thinking of adding https://stefanbirkner.github.io/system-rules/index.html#RestoreSystemProperties

Comment (famod) [reply maven/impl-maven/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/aether/ClassPathScanningTestCase.java]: That's a good idea! Actually, I have implemented such a rule myself in one of my employer's projects.
I'll have a look at `RestoreSystemProperties`.

Comment (famod) [reply maven/impl-maven/src/test/java/org/jboss/shrinkwrap/resolver/impl/maven/aether/ClassPathScanningTestCase.java]: Done!

**Comment by famod** [comment]
Ah, shoot! That dependency was already available. Will fix that...

**Comment by famod** [comment]
> Ah, shoot! That dependency was already available. Will fix that...

Done. I decided to add the dependency to the root `dependencyManagement` since it is used in two modules now.

**Comment by famod** [comment]
We'll, that's strange. `DistributionStageImplTestCase` fails with `system-rules` >= 1.18.0 because it finds multiple matches (but only with JDK7!):

image

This second line is not there with `system-rules` <= 1.17.0 (and generally with JDK8!).

I will stop investigating at this point because it's out of scope for this PR.

@bartoszmajsak Do you want me to squash anything? If not I'm done (literally).

**Comment by famod** [comment]
@bartoszmajsak Thanks for merging. Are you still planning to cut a relase this year?

**Comment by bartoszmajsak** [comment]
> @bartoszmajsak Thanks for merging. Are you still planning to cut a relase this year?

@famod It's been released. I will push to master with the tag as soon as it lands in maven central. Again, many thanks for bringing it back to life :)
